src/test/java/io/vertx/ext/amqp/SenderTest.java

```diff
@@ -334,4 +338,58 @@ private void doSenderClosedRemotelyCallsExceptionHandlerTestImpl(TestContext con
       server.close();
     }
   }
+
+  @Test(timeout = 20000)
+  public void testDynamicSenderWithOptions(TestContext context) throws ExecutionException, InterruptedException {
```

Comment:
Ah. Though my comment applies equally well to the sender case also, it was really the receiver case (since that's the commonly used variant rather than dynamic sender) which I meant to suggest would be better covered by a ProtonServer based test verifying handling of the dynamic address, like the one I then did actually link to. Oops :)

Reply:
I've added a test for the dynamic receiver case too.

Reply:
The new test doesn't validate that the address returned from the dynamic receiver is actually the one it sent from the server, it just checks non-null.

It also seems to takes nearly 11 seconds to complete, when I'd expect 0.x. From looking at the CI job, the overall test suite is rather slow which seems likely to be related (it seems slow even considering all the 3-4 sec periods burned on docker usage).

Reply:
I expect the delay here is because the test server doesn't have a connection close handler defined, so I guess the test is waiting on a close that never completes, with the 10 seconds seeming likely to be spent in BareTestBase tearDown.

In other tests in that class I think it might be waiting 10 seconds because the server was shut down after the client close started but before it completed.

Reply:
@gemmellr The delay is due to this:

```
receiver
              .closeHandler(done -> actualHandler.handle(done.mapEmpty()))
              .close();
```

The close handler is not called. Do you know what can be the reason?

Reply:
Which test? In any case, if it didnt fire then it presumably means it didn't actually receive a detach frame with closed=true. As I say, it looks like some of the tests perhaps stop their server a bit aggressively.

Reply:
I fixed most of them, 2 are remaining: 
* `io.vertx.ext.amqp.ReceiverCreditTest#testInitialCreditInfluencedByConsumerBufferSize`
* `io.vertx.ext.amqp.ReceiverCreditTest#testInitialCredit`

I believe that when the server is closed, the receiver cannot be closed afterward.

Reply:
To me it seems like those 2 tests close their test server in the test method, which will sever its connections, before then later closing the client in the teardown.

The latter looks to close the senders+receivers and then upon firing of their handlers proceed to close the underlying connection. It looks like if any of those never complete, neither does the overall process. None of them (inc connection) will ever close over the wire in this case since the connection was disconnected abruptly before the process even started.

From this Id say:
- The clients connection close handler should be accounting for the underlying connection disconnecting without close frame, both before it begins its close process and again after it has begun.
- Closing the senders and receivers in the connection close could probably just be dropped, a detach is implied when closing the connection, and doing so would simplify the whole process.

The other test in that class (which doesn't seem to have anything to do with credit hehe) doesn't shut down its server at all (though presumably the Vertx instance is closed at some point and does so) which will likely be be why it no longer delays after the other fix but these two still do.

Reply:
I'm going to remove the explicit closing of the receiver and sender qhen the connection is closed

Reply:
All fixed now!